Visual Editor - New, streamlined authentication flow

bttf opened the pull request:
This improves the initial auth flow for when users open the Visual Editor for the first time. This PR gets rid of the copy/paste of API credentials step for a completely automated, no-friction flow.

Accompanies https://github.com/growthbook/devtools/pull/32

Changes:
- New endpoint: `GET /visual-editor/key` which **finds or creates** a personal access token with the role `'visualEditor'`
- `OpenVisualEditor` now sends an async message (via `window.postMessage`) containing above key
  - The chrome extension receives this message, returns a response having received it. Upon receiving, `OpenVisualEditor` navigates to the URL with appropriate query params for the visual editor to load
- We allow bypassing the message response check in cases where we don't detect the visual editor ('Proceed anyway' link)

Notes:
- The chrome extension stores its key in persistent `sync` storage, however we `postMessage` from the app with a key every time anyway. This will help mitigate circumstances where key becomes deleted or out of date for whatever reason
- **Deployment Steps** Timing of deployment is important for backwards compatibility:
1. Chrome extension PR merged first https://github.com/growthbook/devtools/pull/34
2. This PR merged 2nd

- For **self-hosted** accounts, the app may remain out-of-date while the chrome extension is updated. In this case, we inform the user that they can **still set the API key manually**, via the chrome extension options. See screenshot below.

### Screen capture

With no prior API key stored in chrome ext. storage
![no-api-keys](https://github.com/growthbook/devtools/assets/2374625/3f56ef70-4bdc-495d-9bc1-684f31080d9b)

Message for when the API key is not transmitted by GB app for whatever reason (e.g. self-hosted app out of date)
<img width="397" alt="Screenshot 2023-07-25 at 7 58 17 PM" src="https://github.com/growthbook/growthbook/assets/2374625/6a1d7ed5-8cc2-4bf6-9b06-b70aa779b8c3">

jdorn commented on packages/front-end/components/OpenVisualEditorLink.tsx:
```diff
@@ -45,49 +61,81 @@ const OpenVisualEditorLink: FC<{
+
+      track("Open visual editor", {
+        source: "visual-editor-ui",
+        status: "success",
```
We might want to change the status here to `bypass` or something.  Ideally we would see very few people hit this code.  This would let us verify that with data and make sure the auto-auth flow is working for people as intended.

bttf replied:
Addressed

jdorn commented on packages/front-end/components/OpenVisualEditorLink.tsx:
```diff
@@ -45,49 +61,81 @@ const OpenVisualEditorLink: FC<{
-            });
-            return false;
-          }
+          e.preventDefault();
```
If `isNavigating` is true, do we want to make `onClick` do nothing?  Right now, it looks like it would fire duplicate `navigate` calls.

jdorn commented on packages/front-end/components/OpenVisualEditorLink.tsx:
```diff
@@ -45,49 +61,81 @@ const OpenVisualEditorLink: FC<{
+        data: key,
+      },
+      window.location.origin
+    );
```
If someone has the extension installed, but it's an old version, it looks like they will be stuck.  The extension dialog won't show, so there's no way to click the "bypass" link.  It will just stay "navigating" forever.  Can we add a simple `setTimeout` here.  If we haven't received a response from the extension within a short period, show a new dialog with a link letting them bypass.

This will happen to 100% of users when we merge this PR since the extension update won't be released yet.

jdorn replied:
Also, there's no error handling here if the async calls like `getVisualEditorKey` fail.  We should probably catch those and do the same thing - set navigating back to false and show a dialog letting them proceed.  We would also need to refactor this a bit since currently, clicking the bypass link re-runs this entire navigate function, including the API calls, so if those are failing it will be a loop.

bttf commented:
@jdorn Addressed your comments

- There wasn't a double-calling of `navigate` nor a loop occurring. `isNavigating` is only used to render a loading spinner.
- There was one issue of setting `isBypass` to `true` but calling `navigate` right after, not giving it time to take effect. This logic has been moved to a `useEffect`

Tested
- Old ext. version, no chrome ext. detected, 'proceed anyway' flow
- Old ext. version, chrome ext. detected, API key fetch failed flow
- Old ext. version, happy path
- New ext. version, no ext. detected, 'proceed anyway' flow
- New ext. version, ext. detected, API key fetch failed flow
- New ext. version, happy path

jdorn commented on packages/front-end/components/OpenVisualEditorLink.tsx:
```diff
@@ -98,6 +104,38 @@ const OpenVisualEditorLink: FC<{
 
       window.location.href = url;
```
Does this need a return statement here?  I'm pretty sure javascript execution continues after changing the href

bttf replied:
🤷🏾 My assumption was no but I can add for safety

bttf replied:
Added

jdorn commented on packages/front-end/components/OpenVisualEditorLink.tsx:
```diff
@@ -98,6 +104,38 @@ const OpenVisualEditorLink: FC<{
+      setIsNavigating(false);
+      setExtensionDialogText(
+        "We were unable to fetch an API key to initialize the Visual Editor. Please try again or contact support."
+      );
```
The CTA for the extension dialog is hardcoded to "View extension".  Does that make sense for this error too?

bttf replied:
Not really

bttf replied:
Addressed

bttf commented:
@jdorn All comments addressed and tested according to above test plan

bttf commented:
I realized we'll need to update the documentation to reflect the new flow - changes incoming.

bttf commented:
I updated the docs to reflect the streamlined flow (which effectively meant removing copy).

I made some minor improvements as well. There is more work to be done there, but leaving that for the next few PRs that are coming for the Visual Editor.

bttf commented:
Updates:
- Cleaned up logic in `OpenVisualEditorLink`; removed unnecessary state vars
- Tested with all `postMessages` (used for both Visual Editor & DevTools) updated with specific host instead of wildcard
<img width="2032" alt="image" src="https://github.com/growthbook/growthbook/assets/2374625/8e4dbb36-a9c8-4bda-92a7-fdcd97ec2363">

bttf commented:
I updated the deployment sequence of this PR and the [chrome extension PR](https://github.com/growthbook/devtools/pull/34), with the chrome extension needing to be merged first

bttf reviewed (commented):
TODO
- [x] add setTimeout backwards compat back in
- [x] ensure we continue passing exp url until ext. is fully rolled out

jdorn commented:
There's already a dateCreated column for api keys, so no need for that to be in the description.  Also, it should probably mention that it's for the visual editor chrome extension.  Otherwise, users might be confused and think they got hacked or something if they see a random personal access token that they didn't explicitly create.

![image](https://github.com/growthbook/growthbook/assets/1087514/b8efc55f-91ef-4c75-80dc-6662a6c985e3)
